docs/source/format/CDataInterface.rst

```diff
@@ -0,0 +1,777 @@
+the children.
+
+Consumers MUST check that the release callback is non-NULL before calling it.
+The release callback being NULL is not an error, it should just be ignored.
+
+In any case, a consumer MUST not try to access the ArrowArray struct anymore
+after calling its release callback -- including any associated data such
+as its children.
+
+Release callback semantics -- for producers
+'''''''''''''''''''''''''''''''''''''''''''
+
+If producers need additional information for lifetime handling (for
+example, a C++ producer may want to use ``shared_ptr`` for array and
+buffer lifetime), they MUST use the :c:member:`ArrowArray.private_data`
+member to locate the required bookkeeping information.
+
+The release callback MUST not assume that the struct will be located
+at the same memory location as when it was originally produced.  The consumer
+is free to move the struct around (see "Movability").
+
+The release callback MUST walk all children arrays (including the optional
+dictionary) and call their own release callbacks.
+
+The release callback MUST free any data area directly owned by the struct
+(such as the buffers and children arrays).
+
+The release callback MUST mark the array as released, by setting
+:c:member:`ArrowArray.format` to NULL.
+
+Additionally, the release callback MUST be idempotent, which is commonly
+achieved by setting itself to NULL.
+
+Below is a good starting point for implementing a release callback, where the
+TODO area must be filled with producer-specific deallocation code:
+
+.. code-block:: c
+
+   static void ReleaseExportedArray(struct ArrowArray* array) {
+     // This should not be called on already released array
+     assert(array->format != NULL);
+
+     // Release children
+     for (int64_t i = 0; i < array->n_children; ++i) {
+       struct ArrowArray* child = array->children[i];
+       if (child->format != NULL && child->release != NULL) {
+         child->release(child);
+         assert(child->format == NULL);
+       }
+     }
+
+     // Release dictionary
+     struct ArrowArray* dict = array->dictionary;
+     if (dict != NULL && dict->format != NULL && dict->release != NULL) {
+       dict->release(dict);
+       assert(dict->format == NULL);
+     }
+
+     // TODO here: release and/or deallocate all data directly owned by
+     // the ArrowArray struct, such as the private_data.
+
+     // Mark array released
+     array->format = NULL;
+   }
+
+
+Movability
+''''''''''
+
+The consumer can *move* the ``ArrowArray`` struct by bitwise copying (or
+shallow member-wise copying).  Then it MUST mark the source struct released
+(see "released array" above for how to do it) but **without** calling the
+release callback.  This ensures that only one live copy of the struct is
+active at any given time and that lifetime is correctly communicated to
+the producer.
+
+It is possible to move a child array, but the parent array MUST be released
```

Comment:
I don't think I understand how this works exactly, I'll try to re-read the rest of the document.  When does release callback get called in this scenario?    If a client isn't supposed to call children release callbacks, does this imply the memory for the non-chosen children still won't be freed?